File: app/i18n/locale.py
import os
from functools import lru_cache


DEFAULT_LOCALE = 'zh-CN'
SUPPORTED_LOCALES = ('zh-CN', 'zh-TW', 'en-US', 'ja-JP')

_LOCALE_ALIASES = {
    'zh': 'zh-CN',
    'zh-cn': 'zh-CN',
    'zh_hans': 'zh-CN',
    'zh-hans': 'zh-CN',
    'zh-tw': 'zh-TW',
    'zh_hant': 'zh-TW',
    'zh-hant': 'zh-TW',
    'zh-hk': 'zh-TW',
    'zh-mo': 'zh-TW',
    'en': 'en-US',
    'en-us': 'en-US',
    'ja': 'ja-JP',
    'ja-jp': 'ja-JP',
}


def normalize_locale(locale: str | None) -> str:
    if not locale:
        return get_default_locale()

    normalized = locale.strip().replace('_', '-')
    if not normalized:
        return get_default_locale()

    lowered = normalized.lower()
    if lowered in _LOCALE_ALIASES:
        return _LOCALE_ALIASES[lowered]

    for supported in SUPPORTED_LOCALES:
        if lowered == supported.lower():
            return supported

    language = lowered.split('-', 1)[0]
    return _LOCALE_ALIASES.get(language, get_default_locale())


@lru_cache(maxsize=1)
def get_default_locale() -> str:
    return normalize_locale_from_env(os.getenv('DEFAULT_LOCALE'))


def normalize_locale_from_env(locale: str | None) -> str:
    if not locale:
        return DEFAULT_LOCALE

    normalized = locale.strip().replace('_', '-')
    if not normalized:
        return DEFAULT_LOCALE

    lowered = normalized.lower()
    if lowered in _LOCALE_ALIASES:
        return _LOCALE_ALIASES[lowered]

    for supported in SUPPORTED_LOCALES:
        if lowered == supported.lower():
            return supported

    return DEFAULT_LOCALE

File: frontend/src/i18n/locale.ts
export const DEFAULT_LOCALE = 'zh-CN' as const;
export const LOCALE_STORAGE_KEY = 'locale' as const;

export const APP_LOCALES = ['zh-CN', 'zh-TW', 'en-US', 'ja-JP'] as const;

export type AppLocale = (typeof APP_LOCALES)[number];

export function normalizeLocale(value?: string | null): AppLocale {
    if (typeof value !== 'string') {
        return DEFAULT_LOCALE;
    }

    const normalized = value.trim().replace('_', '-').toLowerCase();
    if (!normalized) {
        return DEFAULT_LOCALE;
    }

    if (normalized === 'zh-tw' || normalized === 'zh-hk' || normalized === 'zh-mo'
        || normalized === 'zh-hant' || normalized.startsWith('zh-hant-')) {
        return 'zh-TW';
    }

    if (normalized === 'zh' || normalized === 'zh-cn' || normalized === 'zh-hans'
        || normalized.startsWith('zh-cn-') || normalized.startsWith('zh-hans-')) {
        return 'zh-CN';
    }

    if (normalized === 'ja' || normalized === 'ja-jp' || normalized.startsWith('ja-')) {
        return 'ja-JP';
    }

    if (normalized === 'en' || normalized === 'en-us' || normalized.startsWith('en-')) {
        return 'en-US';
    }

    return DEFAULT_LOCALE;
}

export function toRequestLanguage(locale: AppLocale): AppLocale {
    return locale;
}
